File: smarthomelibary/shlibrary/library/src/main/java/ru/smarthome/library/BaseController.java
package ru.smarthome.library;

import com.google.gson.annotations.Expose;

/**
 * Basic controller interface
 * <br>All controllers must implement this interface
 */
public class BaseController {
    @Expose public ControllerType type;
    @Expose public long guid;
    @Expose public String state;

    public void setNewState(String newState) {}// todo is "String" enough for state?
}

File: smarthome_libraries/common/src/main/java/smarthome/library/common/BaseController.kt
package smarthome.library.common

/**
 * Basic controller interface
 * <br></br>All controllers must implement this interface
 */
open class BaseController(var name: String,
                          var state: ControllerState? = null,
                          var serveState: ControllerServeState = ControllerServeState.IDLE,
                          @Deprecated("use id instead") val guid: Long = 1L,
                          val id: Id = Id(name)) {


    override fun hashCode() = id.hashCode()

    override fun equals(other: Any?) = (other as? BaseController)?.id == id
}